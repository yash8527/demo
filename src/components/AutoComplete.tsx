import React, { useState } from 'react';
import useAutoComplete from '../hooks/useAutoComplete';
import '../styles/AutoComplete.css';
import { AutoCompleteProps } from '../types';

const AutoComplete: React.FC<AutoCompleteProps> = ({ suggestions }) => {
  const { inputValue, setInputValue, filteredSuggestions, isLoading } = useAutoComplete(suggestions);
  const [activeSuggestionIndex, setActiveSuggestionIndex] = useState(0);
  const [showSuggestions, setShowSuggestions] = useState(false);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
    setShowSuggestions(true);
  };

  const handleClick = (suggestion: string) => {
    setInputValue(suggestion);
    setShowSuggestions(false);
  };

  // To handle keyboard navigation and selection of suggestions using arrow keys and Enter key //
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    switch (e.key) {
      case 'Enter':
        setInputValue(filteredSuggestions[activeSuggestionIndex]);
        setShowSuggestions(false);
        break;
      case 'ArrowUp':
        if (activeSuggestionIndex > 0) {
          setActiveSuggestionIndex(activeSuggestionIndex - 1);
        }
        break;
      case 'ArrowDown':
        if (activeSuggestionIndex < filteredSuggestions.length - 1) {
          setActiveSuggestionIndex(activeSuggestionIndex + 1);
        }
        break;
      default:
        break;
    }
  };

  const handleClear = () => {
    setInputValue('');
    setShowSuggestions(false);
  };

  // To highlight the matching text in the suggestions //
  const highlightText = (fullText: string, searchText: string) => {
    const searchRegex = new RegExp(`(${searchText})`, 'gi');
    return (
      <span>
        {fullText.split(searchRegex).map((segment, idx) =>
          searchRegex.test(segment) ? (
            <span key={idx} className="highlight">
              {segment}
            </span>
          ) : (
            segment
          )
        )}
      </span>
    );
  };

  return (
    <div className="autocomplete">
      <div className="input-wrapper">
        <input
          type="text"
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          value={inputValue}
          className="autocomplete-input"
          placeholder='Search for lorem'
          role="combobox"
          aria-expanded={showSuggestions}
        />
        {inputValue && (
          <button
            className="clear-button"
            onClick={handleClear}
            aria-label="Clear input"
            tabIndex={0}
            onKeyDown={(e) => {
              if (e.key === 'Enter') {
                handleClear();
              }
            }}
          >
            &times;
          </button>
        )}
      </div>
      {(isLoading && showSuggestions) && <div className="loader">Loading...</div>}
      {showSuggestions && inputValue && (
        <ul className="suggestions" role="listbox">
          {filteredSuggestions.length ? (
            filteredSuggestions.map((suggestion, index) => (
              <li
                key={suggestion}
                className={index === activeSuggestionIndex ? 'suggestion-active' : ''}
                onClick={() => handleClick(suggestion)}
                role="option"
              >
                {highlightText(suggestion, inputValue)}
              </li>
            ))
          ) : (
            !isLoading && <li className="no-match-found">No match found</li>
          )}
        </ul>
      )}
    </div>
  );
};

export default AutoComplete;