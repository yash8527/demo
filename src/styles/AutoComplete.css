.autocomplete {
    position: relative;
    width: 300px;
  }
  
  .autocomplete-input {
    width: 100%;
    padding: 8px;
    box-sizing: border-box;
  }
  
  .suggestions {
    border: 1px solid #ccc;
    border-top-width: 0;
    list-style: none;
    margin: 0;
    max-height: 600px;
    overflow-y: auto;
    padding: 0;
    position: absolute;
    width: 100%;
    z-index: 1;
    background-color: white;
  }
  
  .suggestions li {
    padding: 8px;
    cursor: pointer;
  }
  
  .suggestions li.suggestion-active {
    background-color: #e9e9e9;
  }
  
  .suggestions li.no-match-found {
    color: #999;
  }
  
  .highlight {
    font-weight: bold;
    background-color: yellow;
  }