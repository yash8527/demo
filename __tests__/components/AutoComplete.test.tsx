import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import AutoComplete from '../../src/components/AutoComplete';
import renderer from 'react-test-renderer';

const suggestions = ['Football', 'Cricket', 'Foosball', 'Basketball'];

describe('AutoComplete Component', () => {
  beforeEach(() => {
    render(<AutoComplete suggestions={suggestions} />);
  });

  it('renders input element', () => {
    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toHaveAttribute('placeholder', 'Search for lorem');
  });

  it('shows "No match found" when no suggestions match', () => {
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'xyz' } });
    expect(screen.getByText(/no match found/i)).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(<AutoComplete suggestions={suggestions} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});