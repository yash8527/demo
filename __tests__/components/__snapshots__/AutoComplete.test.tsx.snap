// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete Component should match snapshot 1`] = `null`;
